routing/routing_helpers.cpp

```diff
@@ -35,5 +42,34 @@ void ReconstructRoute(IDirectionsEngine * engine, IRoadGraph const & graph,
   route.SetTurnInstructions(move(turnsDir));
   route.SetStreetNames(move(streetNames));
   route.SetAltitudes(move(altitudes));
+
+  vector<traffic::SpeedGroup> traffic;
+  if (trafficColoring)
+  {
+    traffic.reserve(routeSegs.size());
+    for (TrafficInfo::RoadSegmentId const & rp : routeSegs)
+    {
+      auto const it = trafficColoring->find(rp);
+      SpeedGroup segTraffic =  (it == trafficColoring->cend()) ? SpeedGroup::Unknown
+                                                               : it->second;
+      // @TODO It's written to compensate an error. The problem is in case of any routing except for osrm
+      // all route points except for begining and ending are duplicated.
+      // See a TODO in BicycleDirectionsEngine::Generate() for details.
+      // At this moment the route looks like:
+      // 0----1    4----5
+      //      2----3    6---7
+      // This route consists of 4 edges and there're 4 items in routeSegs.
+      // But it has 8 items in routeGeometry.
+      // So for segments 0-1 and 1-2 let's set routeSegs[0]
+      // for segments 2-3 and 3-4 let's set routeSegs[1]
+      // for segments 4-5 and 5-7 let's set routeSegs[2]
+      // and for segment 6-7 let's set routeSegs[3]
+      traffic.insert(traffic.end(), {segTraffic, segTraffic});
+    }
+    if (!traffic.empty())
```

Comment:
BicycleDirectionsEngine::Generate() may fill with empty vectors:
|times|, |turns|, |routeGeometry| and |trafficSegs|. This params form an invalid route. 

It's wrong working with routes which have beginning and ending on the same segment. It's better to fix but I guess it's a question of separate PR.

Now I:
- write todo about the issue;
- check (with if) if routeSegs is empty instead of checking |traffic|;
- put CHECK that |traffic| is not empty:
```
  if (engine)
    engine->Generate(graph, path, times, turnsDir, junctions, routeSegs, cancellable);

  // @TODO(bykoianko) If the start and the finish of a route lies on the same road segemnt
  // engine->Generate() fills with empty vectors |times|, |turnsDir|, |junctions| and |routeSegs|.
  // It's not correct and should be fixed. It's necessary to work corrrectly with such routes.

  vector<m2::PointD> routeGeometry;
  JunctionsToPoints(junctions, routeGeometry);
  feature::TAltitudes altitudes;
  JunctionsToAltitudes(junctions, altitudes);

  route.SetGeometry(routeGeometry.begin(), routeGeometry.end());
  route.SetSectionTimes(move(times));
  route.SetTurnInstructions(move(turnsDir));
  route.SetStreetNames(move(streetNames));
  route.SetAltitudes(move(altitudes));

  vector<traffic::SpeedGroup> traffic;
  if (trafficColoring && !routeSegs.empty())
  {
    traffic.reserve(2 * routeSegs.size());
    for (TrafficInfo::RoadSegmentId const & rp : routeSegs)
    {
      auto const it = trafficColoring->find(rp);
      SpeedGroup segTraffic =  (it == trafficColoring->cend()) ? SpeedGroup::Unknown
                                                               : it->second;
      // @TODO It's written to compensate an error. The problem is in case of any routing except for osrm
      // all route points except for begining and ending are duplicated.
      // See a TODO in BicycleDirectionsEngine::Generate() for details.
      // At this moment the route looks like:
      // 0----1    4----5
      //      2----3    6---7
      // This route consists of 4 edges and there're 4 items in routeSegs.
      // But it has 8 items in routeGeometry.
      // So for segments 0-1 and 1-2 let's set routeSegs[0]
      // for segments 2-3 and 3-4 let's set routeSegs[1]
      // for segments 4-5 and 5-7 let's set routeSegs[2]
      // and for segment 6-7 let's set routeSegs[3]
      traffic.insert(traffic.end(), {segTraffic, segTraffic});
    }
    CHECK(!traffic.empty(), ());
    traffic.pop_back();
  }

  route.SetTraffic(move(traffic));
}
```